### ISSUE TITLE
[Bug] create_l1/gather_row matmul path rejects an A-operand with rows % 16 != 0 (no valid_shape equivalent)

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto#2442]
## Background

In pypto-lib, porting the Qwen3-14B decode paged-attention kernel to the
tensor-level `create_l1` / `gather_row` staging path (the path DeepSeek V4's
`qk_pv` uses) has a **compile** problem: the `pl.matmul` A-operand must have a
row count that is a multiple of 16, and GQA decode's M is `q_per_kv` = 5.

Reproduce with:

    python repro_gather_m16.py          # M=5  -> fails
    python repro_gather_m16.py --pass   # M=16 -> compiles

The repro case is **not committed** — save the file below and run it.

<details>
<summary>Reproduction case — <code>repro_gather_m16.py</code> (not committed; save locally to run)</summary>

```python
# Minimal repro: the tensor-level create_l1 / gather_row / matmul path rejects an
# A-operand whose row count is not a multiple of 16.
#
#   python repro_gather_m16.py          -> M=5,  fails in ptoas
#   python repro_gather_m16.py --pass   -> M=16, compiles
#
# GQA decode attention has M = q_per_kv, which is 5 for Qwen3-14B (40 q heads /
# 8 kv heads), so this shape is not a corner case for that model -- it is the
# shape. The tile-level path expresses it with pl.load(..., valid_shape=[M, K]);
# the tensor/gather path has no A-operand equivalent.

import sys

import pypto.language as pl
from pypto.runtime import RunConfig

K = 128
N = 256  # keeps the bf16 B-operand at exactly the 64 KiB L0B limit
ROWS = 64
M = 16 if "--pass" in sys.argv else 5


@pl.jit
def gather_matmul(
    a: pl.Tensor[[ROWS, K], pl.BF16],
    pool: pl.Tensor[[1024, K], pl.BF16],
    out: pl.Out[pl.Tensor[[ROWS, N], pl.FP32]],
) -> pl.Tensor[[ROWS, N], pl.FP32]:
    with pl.spmd(1, name_hint="gather_matmul"):
        _blk = pl.tile.get_block_idx()
        b_l1 = pl.create_l1([N, K], pl.BF16)
        b_l1 = pl.gather_row(b_l1, pool, [0, 0], [0, 0], [N, K])
        acc = pl.matmul(a[0:M, :], b_l1, b_trans=True, out_dtype=pl.FP32)
        out[0:M, :] = acc
    return out


if __name__ == "__main__":
    import torch

    print(f"M = {M}")
    gather_matmul.compile(
        torch.zeros([ROWS, K], dtype=torch.bfloat16),
        torch.zeros([1024, K], dtype=torch.bfloat16),
        torch.zeros([ROWS, N], dtype=torch.float32),
        config=RunConfig(platform="a2a3sim"),
    )
    print("compiled OK")
```

</details>

Reproduction environment:

| Component | Version |
|---|---|
| pypto-lib | `48de053` (branch: `fix/replace-compile-for-test-with-lower`) |
| pypto | `8eab889e` (branch: `feat/mix-kernel-odd-split`) |
| simpler | `1f27a157` (detached) |
| ptoas | `0.57` |
| pto-isa | `f51c92f6` |
| CANN | `9.0.0` |

All four toolchain components match their pins (simpler gitlink, `pto_isa.pin`,
`toolchain/versions.env`).

Diagnosis: **pypto** — the message is emitted by ptoas, but ptoas is enforcing a
real 16-row inner granularity. pypto emits a `pto.alloc_tile` with 5 rows on
this path. On the tile-level path pypto already expresses exactly this case as
`pl.load(..., valid_shape=[M, K])`; the tensor/gather path has no A-operand
equivalent.

### Description

`pl.matmul` whose B-operand comes from `pl.create_l1` requires Tensor operands,
so the A-operand has to be a tensor slice. When that slice's row count is not a
multiple of 16, compilation fails:

```
error: 'pto.alloc_tile' op expects result boxed tile rows to be a multiple of
innerRows (16), but got 5
```

Slicing 16 rows compiles, but in GQA decode that reads past the current kv
head's q rows into the next head's.

This blocks a measured win. On Qwen3-14B decode attention (B16/S2048, in-core
simulator, same shape both sides), staging a 512-token stack into ONE L1 buffer
with four `[128,128]` `gather_row` calls — instead of four separate `pl.load` +
matmul pairs — gives:

| | span | L1->L0 (`LOAD_2D`) | `SET_FLAG` | MTE2 cycles |
|---|---|---|---|---|
| four separate loads | 28.25us | 128 | 200 | 132334 |
| one L1 buffer + gather | **22.37us** | 64 | 113 | 99601 |

**-20.8%**, and compile perf-hints drop from 21 to 9 (the score store becomes one
2048B-row write instead of four 256B-row writes). GQA decode cannot take this
path today.

### Steps to Reproduce

1. Save the file above.
2. `python repro_gather_m16.py` -> fails with the `alloc_tile` error.
3. `python repro_gather_m16.py --pass` -> identical kernel with M=16, compiles.

### Expected Behavior

Either a `valid_shape` (or M-remainder) on the tensor/gather path, matching what
`pl.load` already offers, or padding the `alloc_tile` to 16 rows internally while
keeping the logical M.

### Host Platform

`Linux (aarch64)`

### Additional Context

Related: pypto#594 (automatic `valid_shape` inference), pypto#1031
(odd-dimension split via padding + `valid_shape`), pypto#1956 (first-class
`alloc_tile` IR op) — all adjacent, none covers the tensor/gather matmul
A-operand.

**Comment by lyfne123** [comment]
Companion: pypto#2443 — the two blockers were found together while porting the same kernel, and both must be resolved for a GQA decode attention kernel to use the `create_l1`/`gather_row` staging path.

**Comment by lyfne123** [comment]
Investigated this. Three corrections to the diagnosis, and a fix for the part that is a real defect.

## 1. It is not the `create_l1` / `gather_row` path

Any tensor-level `pl.matmul` fails, with a plain GM `b` and no gather anywhere. Same kernel, only the sizes change:

| M | N | K | result |
|---|---|---|---|
| 5 | 256 | 128 | ❌ `boxed tile rows ... multiple of innerRows (16), but got 5` |
| 8 | 256 | 128 | ❌ same, `got 8` |
| 16 | **5** | 128 | ❌ same, `got 5` — this one is the *B* operand |
| 16 | 256 | **100** | ❌ `boxed tile cols ... multiple of innerCols (16), but got 100` |
| 32 | 256 | 128 | ✅ |

So the constraint is symmetric — every boxed (Mat / L0) tile needs both extents 16-aligned — and it applies to whichever operand happens to be unaligned. `gather_row` was incidental.

In the failing `.pto` all three boxed allocations carry the raw logical extent:

```mlir
%t__tile      = pto.alloc_tile ... : !pto.tile_buf<loc=mat,  rows=5, cols=128, ... fractal=512>
%t__tile_Left = pto.alloc_tile ... : !pto.tile_buf<loc=left, rows=5, cols=128, ... fractal=512>
%acc__tile    = pto.alloc_tile ... : !pto.tile_buf<loc=acc,  rows=5, cols=256, ... fractal=1024>
```

`valid_row = 5` is already correct on each; what is wrong is `rows` in the *type*.

## 2. The tensor/gather path does have an A-operand equivalent

`tensor.slice` has taken a `valid_shape` all along, and `clamp=True` sanctions the window that overhangs the last kv head:

```python
a_pad = pl.slice(q, [16, K], [head * 5, 0], valid_shape=[5, K], clamp=True)
acc   = pl.matmul(a_pad, b_l1, b_trans=True, out_dtype=pl.FP32)
```

That is the exact tensor-level spelling of `pl.load(..., valid_shape=[M, K])`, and it lowers identically: L1 / L0A / L0C all 16 rows with `valid_row = 5`, the DMA transfers 5 rows (`partition_tensor_view<5x128xbf16>`), the store writes 5 rows. Nothing reads past the head.

## 3. The actual defect: `tensor.matmul` dropped the `valid_shape`

`DeduceTensorMatMulType` built its result as `TensorType(output_shape, out_dtype)` — no `tensor_view_` — while `tile.matmul` propagates `{lhs_valid[0], rhs_valid[1]}` (`DeduceMatmulProductInfo`). So the padded operand above produced a product typed `[16, 256]` *fully valid*, and the write was rejected:

```
Subscript-write shape mismatch on source axis 0: window expects 5 elements, source has 16
```

even though the parser explicitly supports this shape — its own comment reads *"a narrowed source (static_shape padded for ISA alignment, valid_shape carrying the logical extent — same pattern `pl.store` accepts on the tile path) is allowed"*. The only reason it did not apply is that the extent never reached the product type. `out[0:16, :]` is not a way out either: in GQA the score tensor has exactly `q_per_kv` rows.

This is also inconsistent within the tensor layer — the elementwise ops already propagate (`preserve_matching_valid_shape` / `MakeFreshTensorType`); matmul was the outlier.

## Fix

`tensor.matmul` / `tensor.matmul_acc` now deduce the valid rectangle the same way their tile counterparts do:

- `tensor.matmul` → `[lhs valid M, rhs valid N]`, each taken from the axis `a_trans` / `b_trans` already selects for the physical shape. Mat-vec / vec-mat carry their single surviving extent; the 0-D dot and the ND batched form stay fully valid, the latter matching its `tile.batch_matmul` lowering.
- `tensor.matmul_acc` → the accumulator's valid region, since the product is written into it.

A valid shape equal to the physical shape canonicalizes away in the `TensorType` constructor, so fully-valid operands deduce exactly the type they did before — no churn.

With that, the kernel in this issue compiles with no `pl.set_validshape` and no widened store:

```python
b_l1  = pl.gather_row(b_l1, pool, [0, 0], [0, 0], [N, K])
a_pad = pl.slice(a, [16, K], [0, 0], valid_shape=[M, K], clamp=True)
out[0:M, :] = pl.matmul(a_pad, b_l1, b_trans=True, out_dtype=pl.FP32)   # M = 5
```

Verified against ptoas 0.57 / a2a3sim. The generated `.pto` matches the `pl.set_validshape` workaround line for line, minus the one `pto.set_validshape` the workaround had to emit.

## What this does not cover

Two things are worth splitting out rather than folding in here:

1. **Auto-padding** (the second option under "Expected Behavior"): rounding the Mat load's declared tile shape up to `BackendHandler::GetL0FractalAlignment()` inside `ConvertTensorToTileOps`, so a bare `a[0:5, :]` works with no `valid_shape` at the call site. It has to happen at IR level, before `AllocateMemoryAddr` charges the tile's bytes. This is #594 territory for the tensor/matmul path.
2. **The diagnosis quality.** `AutoTileMatmulL0` pre-checks `K % align_k` and returns early with `PH-AT-007`, but has no matching pre-check for `M >= min_m`. With M = 5 it calls `ChooseL0Tile` anyway, catches the resulting `ValueError`, and re-emits its text verbatim as a perf hint — so the user is shown an internal assertion aimed at pass authors:

   ```
   [perf_hint PH-AT-005] AutoTileMatmulL0: tile.matmul: ChooseL0Tile rejected configuration — left untouched.
   ChooseL0Tile: allow_padding=false but M=5 is below the cube minimum tile dimension 16. The pass should
   skip this matmul with a perf hint instead of calling the chooser. Check failed: cfg.M >= cfg.min_m at
   .../l0_tile_chooser.cpp:556
   ```

   The matmul is then left untiled, the unaligned `alloc_tile` reaches ptoas, and nothing in between turns the boxing rule into a PyPTO-level error.

Happy to file both as separate issues. Suggest retitling this one to drop `create_l1/gather_row`, since the scope is every tensor-level matmul.
